### PR TITLE
A version of the sync macro that throws earlier

### PR DESCRIPTION
**Pull request by Keno**
I've been looking at what causes deadlocks in our test suite in an effort to cut down on the number of failed tests on CI that result in hangs (since those are hard to diagnose and resolve). I found that by playing with various resource limits, it is easy to create hangs in the test suite. The reason we get a hang rather than a more easily diagnosable error is two fold. We either:
1) Aren't watching for the error (e.g. a socket remote end closing)
2) We aren't propagating the error to the top level

A very common situation for case 2) is that the test is wrapped in `@sync` which doesn't return until all tasks have finished or error'ed. However, in many cases one of the tasks produces data for the others, so if that task errors, the remaining tasks will wait forever. This PR aims to address that situation by introducing a new `@syncany` macro that immediately rethrows any errors thrown by a contained task rather than waiting for all of them to finish. The implementation isn't super performant (it allocates a new task per object being waited on), but should be sufficient for use in the test suite. A better implementation would create a new scheduler object that can be inserted into multiple wait queues.

Example usage of the new macro:

```
@sync begin
    @async error("Hello")
    @async sleep(1000)
end # Waits 1000s

@syncany begin
    @async error("Hello")
    @async sleep(1000)
end # Throws immediately
```

The macro doesn't do any sort of cleanup for the tasks that do not finish, and just lets them run. In the future, we may want to automatically cancel those tasks, but that seemed like a bigger design problem than the simple thing that I wanted (something that propagates error messages more readily, so we see them in the logs).

### CODE REVIEW

**Comment by jonas-schulze** [comment]
Couldn’t this be the new default behavior?

File: base/task.jl [hunk 1]
```diff
@@ -338,6 +364,28 @@ macro sync(block)
...
+"""
+    @syncany
+
+Wait until all lexically-enclosed uses of `@async`, `@spawn`, `@spawnat` and `@distributed`
+are complete, or at least one of them has errored. The first exception is immediately
+rethrown. It is the responsibility of the user to cancel any still-running operations
+during error handling.
+
+!!! Note
+    This interface is experimental and subject to change or removal without notice.
```
Comment (tkf) [inline base/task.jl]: :+1: for marking this experimental. Otherwise, implementing structured concurrency #33248 becomes much harder.

**Comment by vtjnash** [comment]
Maybe we should call this `Base.Future.@async` (with the idea that once optimized, it should be the default)? The `any` name seems odd (because it is waiting for `all`, it’s just early error).

**Comment by Keno** [comment]
You mean `Base.Future.@sync`? I'd be fine with that.

**Comment by tkf** [comment]
Isn't it strange to put an experimental API in `Future`?  I think it makes harder to fix the bigger design problem.

**Comment by c42f** [comment]
Letting tasks drop out the bottom seems like a pretty big change to the semantics of `@sync` so I'm not sure it makes sense to declare it the `Future` without trying it out for a while. I'd also be happier not to export it for the moment if it's experimental.

As you say, it would ideally interact with some kind of cancellation. That's hard but it feels a lot more consistent with the way `@sync` works right now.

It seems handy though and could be a big practical improvement for the test suite. For the moment, could we just name it clearly and not export it?  How about `@sync_failfast` or something equally ugly but explicit?

**Comment by Keno** [comment]
From triage: let's call this `Base.Experimental.@sync`

**Comment by tkf** [comment]
`Base.Experimental.@sync` doesn't sound like a stop-gap measure. Shouldn't the macro name (`@sync`) clarify that:

* this macro has undesirable property (tasks leak),
* it is included temporary only because a real solution is impossible to implement with the current infrastructure, and
* it will be removed or at least will be discouraged to use in the future?

**Comment by Keno** [comment]
Let's merge this and see if it helps CI. Being in experimental makes it explicitly unsupported, so we can change or remove it if we feel like it at any point.

**Pull request merged**